dask/utils.py

```diff
@@ -37,6 +37,61 @@ def apply(func, args, kwargs=None):
         return func(*args)
 
 
+def _deprecated(
+    *,
+    version: str = None,
+    msg: str = None,
+    use_instead: str = None,
+    category: Type[Warning] = FutureWarning,
+):
+    """Decorator to mark a function as deprecated
+
+    Parameters
+    ----------
+    version : str, optional
+        Version of Dask in which the function was deprecated.
+        If specified, the version will be included in the default
+        warning message.
+    msg : str, optional
+        Custom warning message to raise.
+    use_instead : str, optional
+        Name of function to use in place of the deprecated function.
+        If specified, this will be included in the default warning
+        message.
+    category : type[Warning], optional
+        Type of warning to raise. Defaults to ``FutureWarning``.
+
+    Examples
+    --------
+
+    >>> from dask.utils import _deprecated
+    >>> @_deprecated(version="X.Y.Z", use_instead="bar")
+    ... def foo():
+    ...     return "baz"
+    """
+
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            nonlocal version, msg
```

Comment:
I went with the `nonlocal` approach to avoid

```python
UnboundLocalError: local variable 'msg' referenced before assignment
```

Do you have a suggestion for how this can be avoided?

Reply:
I went with the `nonlocal` approach to avoid:

```python
UnboundLocalError: local variable 'msg' referenced before assignment
```

Do you have any suggestions on how this can be avoided?

Reply:
I think it's simplest and cleanest to use a different variable name

Reply:
> this will misbehave if you reuse the decorator

Could you elaborate a little more on how reusing the decorator will lead to issues? Trying our your snippet locally, it looks like reusing the decorator works as expected (though I could be missing something): 

```python
In [1]: from dask.utils import _deprecated

In [2]: d = _deprecated(msg="asdf")
   ...: @d
   ...: def foo(): pass
   ...: @d
   ...: def bar(): pass

In [3]: foo()
<ipython-input-3-c19b6d9633cf>:1: FutureWarning: asdf
  foo()

In [4]: bar()
<ipython-input-4-226d01f48125>:1: FutureWarning: asdf
  bar()
```

Reply:
The problem is when you omit msg:
```python

In [1]: from dask.utils import _deprecated

In [2]: d =  _deprecated()

In [3]: @d
   ...: def foo(): pass

In [4]: @d
   ...: def bar(): pass

In [5]: foo()
<ipython-input-8-c19b6d9633cf>:1: FutureWarning: foo is deprecated and will be removed in a future release.
  foo()

In [6]: bar()
<ipython-input-9-226d01f48125>:1: FutureWarning: foo is deprecated and will be removed in a future release.
  bar()
```
the message of bar incorrectly mentions foo.

Reply:
Ah, I see. Just pushed up some updates which address this